curp/src/client.rs

```diff
@@ -1,30 +1,198 @@
+use itertools::Itertools;
 use parking_lot::RwLock;
 use tokio::time::timeout;
 use tracing::{debug, instrument, warn};
 use utils::{config::ClientTimeout, parking_lot_lock::RwLockMap};
 
 use crate::{
     cmd::{Command, ProposeId},
-    error::{CommandProposeError, CommandSyncError, ProposeError, RpcError, SyncError},
+    error::{
+        ClientBuildError, CommandProposeError, CommandSyncError, ProposeError, RpcError, SyncError,
+    },
     rpc::{
-        self, connect::ConnectApi, FetchLeaderRequest, FetchReadStateRequest, ProposeRequest,
+        self, connect::ConnectApi, protocol_client::ProtocolClient, FetchClusterRequest,
+        FetchClusterResponse, FetchLeaderRequest, FetchReadStateRequest, ProposeRequest,
         ReadState as PbReadState, SyncResult, WaitSyncedRequest,
     },
     LogIndex, ServerId,
 };
 
+/// Client builder
+#[derive(Debug)]
+pub struct Builder<C> {
+    /// Server addresses
+    addrs: Vec<String>,
+    /// Server addresses and ids (used in an inner client)
+    all_members: HashMap<ServerId, String>,
+    /// Local server id (used in an inner client)
+    local_server_id: Option<ServerId>,
+    /// Client timeout
+    timeout: Option<ClientTimeout>,
+    /// Phantom data
+    phantom: PhantomData<C>,
+}
+
+impl<C> Default for Builder<C> {
+    #[inline]
+    #[must_use]
+    fn default() -> Self {
+        Self {
+            addrs: Vec::new(),
+            all_members: HashMap::new(),
+            local_server_id: None,
+            timeout: None,
+            phantom: PhantomData,
+        }
+    }
+}
+
+impl<C> Builder<C>
+where
+    C: Command + 'static,
+{
+    /// Create new client builder
+    #[inline]
+    #[must_use]
+    pub fn new() -> Self {
+        Self::default()
+    }
+
+    /// Set addresses of servers
+    /// # Panics
+    /// Panics when set `all_members` before
+    #[inline]
+    #[must_use]
+    pub fn addrs(self, addrs: Vec<String>) -> Self {
+        Self { addrs, ..self }
+    }
+
+    /// Set server id and all addresses of servers
+    /// # Panics
+    /// Panics when set `addrs` before
+    #[inline]
+    #[must_use]
+    pub fn all_members(self, all_members: HashMap<ServerId, String>) -> Self {
+        Self {
+            all_members,
+            ..self
+        }
+    }
+
+    /// Set client timeout
+    #[inline]
+    #[must_use]
+    pub fn timeout(self, timeout: ClientTimeout) -> Self {
+        Self {
+            timeout: Some(timeout),
+            ..self
+        }
+    }
+
+    /// Set local server id. (only used in an inner client)
+    #[inline]
+    #[must_use]
+    pub fn local_server_id(self, local_server_id: ServerId) -> Self {
+        Self {
+            local_server_id: Some(local_server_id),
+            ..self
+        }
+    }
+
+    /// Build client
+    /// # Errors
+    /// Return error when meet rpc error or missing some arguments
+    #[inline]
+    pub async fn build(self) -> Result<Client<C>, ClientBuildError> {
```

Comment:
We can use `derive_builder` crate to reduce some boilerplate code.

Reply:
`derive_builder` does not support custom build method

Reply:
FYI: https://docs.rs/derive_builder/0.12.0/derive_builder/#build-method-customization

Reply:
FYI: https://docs.rs/derive_builder/0.12.0/derive_builder/#build-method-customization

Reply:
Fixed